Add docs for load_file operator (#606)
astronomer/astro-sdk: docs/astro/sql/operators/load_file.rst

======================================
load_file operator
======================================

When to Use load_file operator
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> 

Reviewer: It is probably worth referencing/include here what we support:
- databases
- file locations
- file types

Author: Done.

There are two use cases of the load_file operator.

> #. When we want a file to be loaded into a database table

Reviewer: We should explain that we also support matching a group of files, and if there are limitations on our pattern matching at the moment (differences between local/gcs/s3).

Author: Done.

#. When we want a file to be converted to dataframe

Case 1: Load into a database table
    you need to pass the `output_table` param to the load_file operator to convert it to a table and the operator returns an instance of the table object passed in `output_table`. The output table will be created if it doesn't exist and will be replaced if it does. We can change this behavior with the `if_exists` parameter.

    .. literalinclude:: ../../../../example_dags/example_load_file.py
       :language: python
       :start-after: [load_file_example_1_start]
       :end-before: [load_file_example_1_end]

Case 2: Load into pandas dataframe
    If you don't pass the `output_table` to the load_file operator it converts the file into a pandas dataframe and returns the reference to dataframe.

    .. literalinclude:: ../../../../example_dags/example_load_file.py
       :language: python
       :start-after: [load_file_example_2_start]
       :end-before: [load_file_example_2_end]

Parameters to use when loading a file to the database table
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> #. **ndjson_normalize_sep** - This parameter is useful when the input file type is NDJSON. Since NDJSON file can be multidimensional, we normalize the data to two-dimensional data, so that it is suitable to be loaded into a table and this parameter is used as a delimiter for combining columns names if required.

Reviewer: Priority-wise, it feels this should be the last option to be listed - since it is NDJSON specific.
We probably want to have `if_exists` and `output_table` at the top of the arguments list.

Author: Done.

    example:
        input JSON:

        .. code:: python

           {"a": {"b": "c"}}

        output table:

        .. list-table::
           :widths: auto

           * - a_b
           * - c

    Note - columns a and b are merged to form one col a_b and `_` is used as a delimiter.

    .. literalinclude:: ../../../../example_dags/example_load_file.py
       :language: python
       :start-after: [load_file_example_3_start]
       :end-before: [load_file_example_3_end]

> #. **if_exists** - This parameter comes in handy when the table you trying to create already exists. In such a case, we have two options, either replace or append. Which can we accomplish by passing ``if_exists='append'`` or ``if_exists='replace'``.

Reviewer: It is essential to document that if `replace` is chosen, we'll be dropping the table and replacing it with a new table. This new table column types will be defined either by the `table.columns` property, if the user sets that, or by automatically inferring the column types using Pandas using up to N rows (check the property name) from the first file of the given path.

Author: Done.

    .. literalinclude:: ../../../../example_dags/example_load_file.py
       :language: python
       :start-after: [load_file_example_4_start]
       :end-before: [load_file_example_4_end]

#. **output_table** - We can specify the output table to be created by passing in this parameter, which is expected to be an instance of ``astro.sql.table.Table``. Users can specify the schema of tables by passing in the ``columns`` parameter of ``astro.sql.table.Table`` object, which is expected to be a list of the instance of ``sqlalchemy.Column``. If the user doesn't specify the schema the schema is inferred using pandas.

    .. literalinclude:: ../../../../example_dags/example_load_file.py
       :language: python
       :start-after: [load_file_example_5_start]
       :end-before: [load_file_example_5_end]

Parameters to use when loading a file to pandas dataframe
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> #. ``columns_names_capitalization`` Control col names case of the dataframe generated from the file. The default value is ``original``.

Reviewer: Since this is also used in the case of Snowflake's table creation (and affects the table creation in any database), perhaps we should not keep it in the dataframe-specific section.

Reviewer: I feel that this argument will not work if the user specifies the `table.columns` and `if_exists=replace`. Could you confirm and log a ticket if that is the case?

Author: Added `columns_names_capitalization` to database section as well as dataframe, since it has a different meanings in both cases.

        *  **original** - Remains the same as the input file
        *  **upper** - Convert to uppercase
        *  **lower** - Convert to lowercase

    .. literalinclude:: ../../../../example_dags/example_load_file.py
       :language: python
       :start-after: [load_file_example_6_start]
       :end-before: [load_file_example_6_end]

> How load_file Works

Author: @tatiana I agree, I'm waiting on the following PR to add the concepts page, will move the internal explanation to a concept page.
https://github.com/astronomer/astro-sdk/pull/609

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> .. figure:: ../../../images/defaultPath.png

Reviewer: It may be tidier to have this diagram horizontal:
```
() -> () -> ()
```

The same applies to the second diagram. It would be great if you could also document how this diagram was generated, and if possible give means of other team members updating them if necessary.

Author: @tatiana Where should I this info? in PR or doc itself?

When we load a file located in cloud storage to a cloud database, internally the steps involved are:

Steps:

#. Get the file data in chunks from file storage to the worker node.
#. Send data to the cloud database from the worker node.

This is the default way of loading data into a table. There are performance bottlenecks because of limitations of memory, processing power, and internet bandwidth of worker node.

Improving bottlenecks by using native paths
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.. figure:: ../../../images/nativePath.png

Some of the cloud databases like Bigquery and Snowflake support native paths (complete list of supported native paths :ref:`supported_native_path`) to ingest data from cloud storage directly. Using this we can ingest data much quicker and without any involvement of the worker node.

Steps:

#. Request destination database to ingest data from the file source.
#. Database request file source for data.

> This is a faster way to ingest data as there is only one network call involved and usually the bandwidth between vendors is high. Also, there is no requirement for memory/processing power of the worker node, since data never gets on the node. There is significant performance improvement due to native paths as evident from benchmarking results.

Reviewer: This may not be the case.

Depending on the native load implementation, we have an "expensive" setup (e.g. creating Google Storage Transfer configuration or creating Snowflake remote staging) - which means that for small datasets, the native load may be slower when compared to using the Pandas dataframe approach.

Author: Updated the doc to reflect this fact.

> This is a faster way to ingest data as there is only one network call involved and usually the bandwidth between vendors is high. Also, there is no requirement for memory/processing power of the worker node, since data never gets on the node. There is significant performance improvement due to native paths as evident from benchmarking results.

Reviewer: We're missing a link to the benchmark results

Author: Added link to benchmark doc

**Note** - By default the native path is enabled and will be used if the source and destination support it, this behavior can be altered by the ``use_native_support`` param.

> Parameters for native path

Reviewer: Please, let remove any occurrences of "native path" by one of the following:
* native support
* native transfer
* native load

Author: Used - native transfer in docs

======================================
load_file operator
======================================

When to use the ``load_file`` operator
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

There are two use cases of the ``load_file`` operator.

#. When we want a (group of) file(s) to be loaded into a database table
#. When we want a (group of) file(s) to be loaded to a Pandas dataframe

Case 1: Load into a database table

>     you need to pass the `output_table` param to the load_file operator to convert it to a table and the operator returns an instance of the table object passed in `output_table`. The output table will be created if it doesn't exist and will be replaced if it does. We can change this behavior with the `if_exists` parameter.

Reviewer 2: ```suggestion
    you need to pass the ``output_table`` param to the load_file operator to convert it to a table and the operator returns an instance of the table object passed in ``output_table``. The output table will be created if it doesn't exist and will be replaced if it does. We can change this behavior with the `if_exists` parameter.
```

rst requires two quotes for inline code blocks

    .. literalinclude:: ../../../../example_dags/example_load_file.py
       :language: python
       :start-after: [START load_file_example_1]
       :end-before: [END load_file_example_1]

Case 2: Load into pandas dataframe

>     If you don't pass the `output_table` to the load_file operator it converts the file into a pandas dataframe and returns the reference to dataframe.

Reviewer 2: ```suggestion
    If you don't pass the ``output_table`` to the load_file operator it converts the file into a pandas dataframe and returns the reference to dataframe.
```

#. **ndjson_normalize_sep** - This parameter is useful when the input file type is NDJSON. Since NDJSON file can be multidimensional, we normalize the data to two-dimensional data, so that it is suitable to be loaded into a table and this parameter is used as a delimiter for combining columns names if required.
    example:
        input JSON:

        .. code:: python

           {"a": {"b": "c"}}

        output table:

        .. list-table::
           :widths: auto

           * - a_b
           * - c

    Note - columns a and b are merged to form one col a_b and `_` is used as a delimiter.

    .. literalinclude:: ../../../../example_dags/example_load_file.py
       :language: python
       :start-after: [START load_file_example_3]
       :end-before: [END load_file_example_3]

.. _table_schema:

> Inferring Table Schema
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
Inferring Table Schema
~~~~~~~~~~~~~~~~~~~~~~
```

There are two ways to get the schema of the table to be created, listed by priority

#. **User specified schema** - Users can specify the schema of the table to be created in the Table object like the ``output_table`` section in :ref:`custom_schema`

#. **Auto schema detection** - if the user doesn't specify the schema in the table object then by using the top 1000 rows of the table we infer the schema of the table. The default value is 1000, which can be changed by creating an environment variable

>     .. code:: shell

Reviewer 2: ```suggestion
    .. code-block:: shell
```

>        Shell

Reviewer 2: this shouldn't be here, right?

       AIRFLOW__ASTRO_SDK_LOAD_TABLE_AUTODETECT_ROWS_COUNT

    or within airflow config

>     .. code:: shell

Reviewer 2: ```suggestion
    .. code-block:: ini
```

       [astro_sdk]
       load_table_autodetect_rows_count = 1000

    Note - this only applies to :ref:`filetype` JSON, NDJSON and CSV, PARQUET have type information and we don't need to infer it.

Parameters to use when loading a file to pandas dataframe
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
#. ``columns_names_capitalization`` Control col names case of the dataframe generated from the file. The default value is ``original``.
        *  **original** - Remains the same as the input file
        *  **upper** - Convert to uppercase
        *  **lower** - Convert to lowercase

    .. literalinclude:: ../../../../example_dags/example_load_file.py
       :language: python
       :start-after: [START load_file_example_6]
       :end-before: [END load_file_example_6]

> Parameters for native transfer
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
Parameters for native transfer
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

        .. literalinclude:: ../../../../example_dags/example_load_file.py
           :language: python
           :start-after: [START load_file_example_8]
           :end-before: [END load_file_example_8]

#. **enable_native_fallback** -  When ``use_native_support`` is True, we try to use the native transfer, and if this fails we try to use the default path to load data, giving the user a warning. If you want to change this behavior pass ``enable_native_fallback=False``.

        .. literalinclude:: ../../../../example_dags/example_load_file.py
           :language: python
           :start-after: [START load_file_example_9]
           :end-before: [END load_file_example_9]

.. _supported_native_path:

Supported native transfers
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.. list-table::
   :widths: auto

   * - File Location
     - Database
     - native_support_kwargs params
   * - S3
     - Bigquery
     - https://cloud.google.com/bigquery-transfer/docs/s3-transfer#bq
   * - GCS
     - Bigquery
     - https://cloud.google.com/bigquery-transfer/docs/cloud-storage-transfer#bq
   * - S3
     - Snowflake

>      -

Reviewer 2: Are there links to it?

   * - GCS
     - Snowflake

>      -

Reviewer 2: For this too?

> Patterns in File path
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
Patterns in File path
~~~~~~~~~~~~~~~~~~~~~
```

Load file can also resolve patterns in file path, there are three types of patterns supported by load file based on the :ref:`file_location`

#. **Local** - On local we support glob pattern - https://docs.python.org/3/library/glob.html
#. **S3** - prefix in file path - https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-prefixes.html
#. **GCS** - prefix and wildcard in file path - https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames

    .. literalinclude:: ../../../../example_dags/example_load_file.py
       :language: python
       :start-after: [START load_file_example_10]
       :end-before: [END load_file_example_10]

> Inferring File Type
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
Inferring File Type
~~~~~~~~~~~~~~~~~~~
```

There are two ways we infer :ref:`filetype` in the following order:

#. **File object** - If the user has passed the ``filetype`` param while declaring the ``astro.files.File`` object, we use that as file type. Valid values are listed :ref:`filetype`

    .. literalinclude:: ../../../../example_dags/example_load_file.py
       :language: python
       :start-after: [START load_file_example_10]
       :end-before: [END load_file_example_10]

    Note - This param becomes mandatory when the file path don't have extension.

#. **From file extensions** - When we create ``astro.files.File`` object and passed a fully qualified path like below, file extensions are used to infer file types. Here the file type is CSV.

    .. literalinclude:: ../../../../example_dags/example_load_file.py
       :language: python
       :start-after: [START load_file_example_4]
       :end-before: [END load_file_example_4]

Note - 1st way take priority over 2nd way.

> Loading data from HTTP API
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
Loading data from HTTP API
~~~~~~~~~~~~~~~~~~~~~~~~~~
```

======================================
load_file operator
======================================

When to Use load_file operator
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

There are two use cases of the load_file operator.

> #. When we want a file to be loaded into a database table
> #. When we want a file to be converted to dataframe

Reviewer: ```suggestion
#. When we want a (group of) file(s) to be loaded into a database table
#. When we want a (group of) file(s) to be converted to dataframe
```

Case 1: Load into a database table

>     you need to pass the `output_table` param to the load_file operator to convert it to a table and the operator returns an instance of the table object passed in `output_table`. The output table will be created if it doesn't exist and will be replaced if it does. We can change this behavior with the `if_exists` parameter.

Reviewer 3: Should we cover the difference between creating a named Table vs. an unnamed one (temp table)?

Reviewer: Great point, @mikeshwe! @utkarsharma2 is covering this in the concepts and cleanup documentation (https://github.com/astronomer/astro-sdk/pull/609). What do you think if we add here a reference to the Table concepts section?

Author: @mikeshwe I thought so too, but I thought it should be on a separate page. Please check the PR  - https://github.com/astronomer/astro-sdk/pull/609

        This feature is enabled by default, to disable it refer to the below code.

        .. literalinclude:: ../../../../example_dags/example_load_file.py
           :language: python
           :start-after: [load_file_example_7_start]
           :end-before: [load_file_example_7_end]

        To check if the native path will be used for data transfer for a combination of file location and database, refer section :ref:`supported_native_path`

#. **native_support_kwargs** - Since we support multiple databases they may require some parameters to process a file or control error rate etc, those parameters can be passed in ``native_support_kwargs``. These parameters are passed to the destination database.

        Check for valid parameters based on **file location** and **database** combination in section :ref:`supported_native_path`

        .. literalinclude:: ../../../../example_dags/example_load_file.py
           :language: python
           :start-after: [load_file_example_8_start]
           :end-before: [load_file_example_8_end]

> #. **enable_native_fallback** -  When ``use_native_support`` is True, we try to use the native path, and if this fails we try to use the default path to load data, giving the user a warning. If you want to change this behavior pass ``enable_native_fallback=False``.

Reviewer 3: Would it be possible to comment on why someone would want to disable native fallback?

Author: Added in the docs.

        .. literalinclude:: ../../../../example_dags/example_load_file.py
           :language: python
           :start-after: [load_file_example_9_start]
           :end-before: [load_file_example_9_end]

.. _supported_native_path:

Supported Native Paths
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.. list-table::
   :widths: auto

   * - File Location
     - Database
     - native_support_kwargs params
   * - S3
     - Bigquery
     - https://cloud.google.com/bigquery-transfer/docs/s3-transfer#bq
   * - GCS
     - Bigquery
     - https://cloud.google.com/bigquery-transfer/docs/cloud-storage-transfer#bq
   * - S3
     - Snowflake
     -
   * - GCS
     - Snowflake
     -

.. _file_location:

Supported File Location
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Users can load file from all the supported file location that are listed below:

.. literalinclude:: ../../../../src/astro/constants.py
   :language: python
   :start-after: [filelocation_start]
   :end-before: [filelocation_end]

.. _filetype:

Supported File Type
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Users can load the file of all the supported file types that are listed below:

Reviewer 3: If someone specifies the file format in File(filetype) parameter, and that contradicts the .suffix of the file name, I'm assuming that the filetype specified wins. Can you confirm and document?

Author: Added to docs

#. **Local** - On local we support glob pattern - https://docs.python.org/3/library/glob.html
#. **S3** - prefix in file path - https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-prefixes.html
#. **GCS** - prefix and wildcard in file path - https://cloud.google.com/storage/docs/gsutil/addlhelp/WildcardNames

    .. literalinclude:: ../../../../example_dags/example_load_file.py
       :language: python
       :start-after: [load_file_example_10_start]
       :end-before: [load_file_example_10_end]

Inferring File Type
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

There are two ways we infer :ref:`filetype`

#. **File object** - If the user has passed the ``filetype`` param while declaring the ``astro.files.File`` object, we use that as file type. Valid values are listed :ref:`filetype`

    .. literalinclude:: ../../../../example_dags/example_load_file.py
       :language: python
       :start-after: [load_file_example_10_start]
       :end-before: [load_file_example_10_end]

    Note - This param becomes mandatory when the file path don't have extension.

> #. **From file extensions** - When we create ``astro.files.File`` object and passed a fully qualified path like below, file extensions are used to infer file types. Here the file type is CSV.

Reviewer 3: Please see comment on line #210 above

Author: Added to docs

======================================
load_file operator
======================================

> When to Use load_file operator

Reviewer: ```suggestion
When to use the ``load_file`` operator
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> There are two use cases of the load_file operator.

Reviewer: ```suggestion
There are two use cases of the ``load_file`` operator.
```

#. When we want a (group of) file(s) to be loaded into a database table

> #. When we want a (group of) file(s) to be converted to dataframe

Reviewer: ```suggestion
#. When we want a (group of) file(s) to be loaded to a Pandas dataframe
```
